Add `api --paginate` option

mislav opened the pull request:
`gh api --paginate repos/:owner/:repo/issues` automatically fetches and prints all pages of all results until there are no more pages.

TODO:
- [x] GraphQL pagination
- [x] automatically set `per_page=100` in pagination mode?

Fixes https://github.com/cli/cli/issues/1184

mislav committed:
api command: add support for REST pagination

mislav committed:
Add assertion for 1st api request before pagination

mislav committed:
api command: add GraphQL support for `--paginate`

vilmibm reviewed (approved):
i love it

vilmibm commented:
>  automatically set per_page=100 in pagination mode?

seems reasonable? why wouldn't we?

AliabbasMerchant commented:
1 small point:  
paginate means "dividing a document into discrete pages, either electronic pages or printed pages."  
Since we are doing the exact opposite, I don't think the command should be named `--paginate`. Some users (like me) will find it unintuitive.  
Maybe `--compile`, or `--depaginate` or `--combine`?

tatar28Filem0963200833 commented:
> > ตั้งค่า per_page = 100 โดยอัตโนมัติในโหมดเลขหน้า?
> #1222 
> ดูเหมือนว่าสมเหตุสมผล
Duplicate of # ทำไมเราไม่

tatar28Filem0963200833 commented:
> `gh api --paginate repos/:owner/:repo/issues` ดึงและพิมพ์หน้าทั้งหมดของผลลัพธ์ทั้งหมดโดยอัตโนมัติจนกว่าจะไม่มีหน้าเพิ่มเติม
> 
> ทำ:
> 
> * [x]  เลขหน้า GraphQL
> * [ ] ตั้งค่าโดยอัตโนมัติ`per_page=100`ในโหมดเลขหน้า?
> 
> แก้ไข [# 1184](https://github.com/cli/cli/issues/1184)

https://github.com/cli/cli/pull/1222#issue-435328723

mislav commented:
@AliabbasMerchant That's a valid point, but I still feel that "paginate" describes this feature the best since it performs "pagination" (a concept described by that term in GitHub API docs) for you.

I feel that none of the alternate terms that you've suggested are established enough that we could use them as alternatives. Furthermore, I think that `--paginate` is useful to keep as a term because the caller needs to be aware that multiple HTTP requests will be made by the client to get all pages of results (and each will count towards the API request limit).

mislav committed:
Automatically add `per_page=100` to paginated REST requests

Most endpoints respect this parameter by default. Those that don't will
just ignore it. The `per_page=100` parameter is not added if there is
already a `per_page` parameter specified in the request.

bigcook12345 commented:
Closed and opened so soon..?
